cfApplication.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+type Application struct {
+	Name      string            `json:"name"`
+	Metadata  Metadata          `json:"metadata,omitempty"`
+	Env       map[string]string `json:"env,omitempty"`
+	Services  Services          `json:"services,omitempty"`
+	Processes Processes         `json:"processes,omitempty"`
+	Replicas  uint              `json:"replicas,omitempty"`
+}
+
+type Metadata struct {
+	Labels      map[string]string `json:"labels,omitempty"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+type Routes []Route
+
+type Route struct {
+	URL      string        `json:"fqdn,omitempty"`
+	Protocol RouteProtocol `json:"protocol,omitempty"`
+}
+
+type RouteSpec struct {
+	Type   RouteType `json:"type"`
+	Routes Routes    `json:"routes"`
+}
+
+type RouteType string
+
+const (
+	Default RouteType = "default"
+	Random  RouteType = "random"
+)
+
+type RouteProtocol string
+
+const (
+	HTTP  RouteProtocol = "http"
+	HTTPS RouteProtocol = "https"
+	TCP   RouteProtocol = "tcp"
+)
+
+// Services represents a slice of Service
+type Services []Service
+
+// Service contains the specification for an existing Cloud Foundry service required by the application
+type Service struct {
+	// Name represents the name of the Cloud Foundry service required by the application
+	Name string `json:"name"`
+	// Parameters contain the k/v relationship for the aplication to connect to the service
+	Parameters map[string]interface{} `json:"parameters,omitempty"`
+}
+
+// Processes represents a slice of Processes
+type Processes []Process
+
+// Process contains the abstraction of the specification of a Cloud Foundry Process
+type Process struct {
+	// Name represents the name of the process
+	Name string `json:"name"`
+	// Image represents the pull spec of the container image
+	Image string `json:"image"`
+	// Amount of memory requested by the process
+	Memory string `json:"memory,omitempty"`
+	// Amount of persistent disk requested by the process
+	Disk string `json:"disk,omitempty"`
+	// LivenessProbeInitialDelay represents the numbef of seconds before start checking the first health of the process.
+	LivenessProbeInitialDelay uint `json:"livenessProbeInitialDelay,omitempty"`
+	// LivenessProbeRetries represents the number of retries to attempt before failing the liveness probe of the process
+	LivenessProbeRetries uint `json:"livenessProbeRetries,omitempty"`
+	// LivenessProbe represents the endpoint location where to perform the liveness probe check
+	LivenessProbe string `json:"livenessProbe,omitempty"`
+	// LivenessProbeTimeout represents the number of seconds in which the liveness probe can be considered as timedout
+	LivenessProbeTimeout uint `json:"livenessProbeTimeout,omitempty"`
```

Comment:
I'm ok with these values, do we want to use a sub structure like `LivenessProbe` with `delay`, `retries, `url`, and `timeout` fileds?

Reply:
We can, just didn't want to start creating sub structures that I think would always be somehow populated. I mean, liveness probes will be required for a pod to run, so I expect them to exist. If these fields were optional from K8s perspective I would have taken the sub structure approach since it would have simplified the check. WDYT?
I did the same with resources (memory and disk), I had them in a sub structure but I reckon they will always have at least the memory provided, but we can move them to a sub struct.

Reply:
At least the `LivenessProbe` field should be required, the others could use the default value in K8s if they're not provided. WDYT?

Reply:
I like the idea of having default values, but how can we map them? will they be part of the Helm chart? Maybe It's a good idea to document the default values in the canonical configuration, but not sure if this phase is proper one

Reply:
Looks like the documentation already defines default values for liveness probe:
https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#health-check-http-ep

And I realize I missed the readiness probe

Reply:
Updated